### DISCUSSION TITLE
Unexpected incorrect reading from Magnetometer during long range test

### DISCUSSION THREAD

**Discussion by alccrl** [iNavFlight/inav#7991] [Show and tell]
Hi everybody,

during a quite challeging test flight, in not very good weather, but under good rssi recetion and quad control,
I noticed that compass heading was incorrect, whilst azimuth was correct.
Here is a screenshot:
![Heading error](https://user-images.githubusercontent.com/78967144/168442821-db8eb3b4-a0c3-4883-bd86-f491e8c95116.png)

It would very difficult to reexecute this test (read below why), so I decided to start this discussion.
I know that compass setup is much sensible to a lot of factors, but why not to try to quantify them, in some way ?

In the screenshot there is a difference of about 90 degrees vs real heading (quad was coming down back to home in acro mode, direction north, more or less, so azimuth is shown 184).

Why this difference between heading (287) and azimuth (184) ?
A few data about flight:
- very high quad speed (101 km/h) 
- humidity in air
- high above sea level altitude
- windy
- descending at fast rate in ACRO mode
- quad was facing forward (analyzed logs) and no deriving for wind, so heading should match azimuth (184+180 => 4 degrees) (or not (?) ...)
System quad setup:
- FC: F7 Mtek STD
- GPS and Compass : M8Q-5883 (centered top of frame, in free air, not flow or box protected)
- Video system at 5.8 GHz Analog
- 6" propellers on EMAX RS2205 motors
- ESCs : FlyColor X-Cross BL32-Slim
- Frame : QAV-250

The question is :
Can humidity or high (because high speed) air flow against compass cause incorrect reading ?
Can compass device or carbon frame be electrified because of strong air flow against it ?

Could this be caused by a big negative quad pitch/attitude (if any, to be analyzed in log) that leads to more compass reading on the vertical axis than horizontal, with less horizontal precision ?
(I do compass calibration alwasy in 3d-mode).

Did anybody encounter similar behaviour/situation ?

Alfredo

**Comment by Jetrell** [comment]
INAV doesn't have the ability to display Magnetometer bearing and GPS heading... It only displays the GPS heading on the OSD.
But it does take the magnetic bearing into account when navigating.

Your heading of 287° NW, is the heading the quad is on, from its perspective.
While the Azimuth is the crafts position, from the Home location, while viewed from a northerly perspective. 
In other words.. If YOU are facing north. And the Azimuth is displaying 184°, then your quad is behind you, to the South of your position.

**Reply by breadoven** [reply]
Not sure that's true. OSD heading just shows attitude.yaw. On a multirotor with a working magnetometer GPS course over ground isn't used at all for the yaw value, just uses the mag value.

**Reply by alccrl** [reply]
I think GPS Heading (course) is not part of OSD parameters pool, I did not find it.
This is why I added Azimuth, that is available on OSD, in order to have both gps relative position (azimuth) and mag heading displayed.
Azimuth does not mind which part of quad is oriented while flying, and mag heading supplies this lack. Since mag heading is used in navigation (to exit hovering, I guess), both should give a sufficient idea, together with altitude and speed, of navigation status on OSD, if both are reliable.
I had a few direct experiences (not completely understood crashes) that leaded me to prefer ACRO mode vs navigation modes, because mag readings may not always be reliable and may be the cause of quad loss of control.
This is why I am practising ACRO a lot with simulator, and I got my good results, let's say it ...it seems to work ...!!

To be honest,
I do not understand why gps course is not available on OSD, all phone outdoor apps have it, so it should not be difficult to be implemented, (it may even be available at NMEA UART level (?)).

Alfredo

**Reply by breadoven** [reply]
GPS course over ground would be useful for a quad given it's reliable at pretty low speeds (3m/s threshold used for INAV). It would certainly show if the magnetometer was misbehaving so long as you were flying forward of course.

However, looking at the screen shot again I think the home direction arrow shows what's happening. The quad in the screen shot is actually heading 287 degs and it is 184 degs (south) of home. The home direction arrow is therefore pointing to the right since that is where home is. The home direction arrow uses GPS position to calculate the direction so it's unaffected by the compass. Unless the GPS data in the log says otherwise it would appear the compass is OK in fact and the quad was simply moving E to W at this point of the flight, not toward home.

**Reply by alccrl** [reply]
I'll better check (it will take me time) the blackbox and RC telemetry logs, because I have no actioncam video (usually I have it), because cam was set into timelapse mode (...:-)...), so high res video is  useless (it is not highres ...:-)) but at first analysis the quad was facing north in the screenshot (flight path attached here).
![Flight summary](https://user-images.githubusercontent.com/78967144/169103844-bb793f83-f256-4220-8f74-6cdb45ad010e.jpg)

Alfredo

**Reply by alccrl** [reply]
Addendum: I always calibrate compass/mag in 3d mode (matek 8 gos+ compass)

**Reply by Jetrell** [reply]
> Not sure that's true. OSD heading just shows attitude.yaw. On a multirotor with a working magnetometer GPS course over ground isn't used at all for the yaw value, just uses the mag value.

The reason I came to this conclusion. Was from a couple of builds, that experienced course/ heading divergence between the quads direction to a target location and the yaw angle it was fly at.. _It was crabbing along sideways to reach the target_.
The log pointed toward Heading being set by the magnetometer, and the course it was traveling over ground, by the GPS....

**Comment by alccrl** [comment]
Thanks,

but ... 
if I am north, and quad is coming toward back me from south, I guess it is facing north, it is not rolling 90 degrees toward me.
Anyway, ... the quad came back, this is most important, after all ... :-).

Alfredo

**Reply by Jetrell** [reply]
As you eluded to... The magnetometer can experience a divergence from its magnetic bearing, not just from declination, but at certain bank angles (depending on the Mag/GPS positioning) if a motors magnet flux intersecting the magnetometer.  

A magnets strongest flux is produce at the motors side... not above or below that axis.
So even if there appears to be no magnetometer interference when the quad is sitting level.. This doesn't mean it won't occur when the quad is banked at certain angles (on both the pitch or roll axis)

**Comment by alccrl** [comment]
Yes,
I think this is what happened.
Luckily, I was using ACRO mode and I am convincing myself that is better to use ACRO mode as soon as it is safe, because it does not depend on magnetometer, that is an unsafe device, at least in "retail" quadcopters.

Alfredo

**Comment by alccrl** [comment]
Many months later .... :

I discovered on my pc a compass module HMC5883L application note, saved in a hurry after I downloaded it from Internet, and not instantly read, may be ...

This appnote states not affordable data for high pith values (quad in divind mode, for instance). Should be used an accellerometer to compensate for. errors... don't know if Inav does use accellerometer for magnetometer value compensations....

[HMC5883L App Note (pag 1).pdf](https://github.com/iNavFlight/inav/files/14474462/HMC5883L.App.Note.pag.1.pdf)

Hope this helps,

Alfredo
